File: 1293-shortest-path-in-a-grid-with-obstacles-elimination/1293-shortest-path-in-a-grid-with-obstacles-elimination.cpp
class Solution {
public:
    int shortestPath(vector<vector<int>>& grid, int k) {
        int n=grid.size();
        int m=grid[0].size();
        queue<pair<int,pair<int,int>>> q;
        q.push({k,{0,0}});
        vector<vector<vector<int>>> vis(n,vector<vector<int>>(m,vector<int>(k+1,0)));
        vis[0][0][k]=1;
        int steps=0;
        while(q.size()>0){
            int size=q.size();
            while(size--){
                int r=q.front().first;
                int x=q.front().second.first;
                int y=q.front().second.second;
                q.pop();

                if(x==n-1 and y==m-1) return steps;
                int dx[]={-1,0,+1,0};
                int dy[]={0,+1,0,-1};
                for(int i=0;i<4;i++){
                    int nrow=x+dx[i];
                    int ncol=y+dy[i];
                    if(nrow>=0 and nrow<n and ncol>=0 and ncol<m){
                        if(grid[nrow][ncol]==0 and vis[nrow][ncol][r]==0){
                            q.push({r,{nrow,ncol}});
                            vis[nrow][ncol][r]=1;
                        }
                        else if(grid[nrow][ncol]==1 and r>0 and vis[nrow][ncol][r-1]==0){
                            q.push({r-1,{nrow,ncol}});
                            vis[nrow][ncol][r-1]=1;
                        }
                    }
                }
            }
            steps++;
        }
        return -1;
    }
};

File: 1293-shortest-path-in-a-grid-with-obstacles-elimination/1293-shortest-path-in-a-grid-with-obstacles-elimination.java
class Pair{
    int steps;
    int row;
    int col;
    int obs;

    Pair(int steps,int row,int col,int obs){
        this.steps=steps;
        this.row=row;
        this.col=col;
        this.obs=obs;
    }
};
class Solution {
    public int shortestPath(int[][] grid, int k) {
        int n=grid.length;
        int m=grid[0].length;
        Queue<Pair> q=new LinkedList<>();
        int[][][] vis=new int[n][m][k+1];
        q.offer(new Pair(0,0,0,k));
        vis[0][0][k]=1;

        while(q.size()>0){
            Pair p=q.poll();
            int currSteps=p.steps;
            int row=p.row;
            int col=p.col;
            int remObs=p.obs;

            if(row==n-1 && col==m-1) return currSteps;
            
            int[] drow=new int[]{0,+1,0,-1};
            int[] dcol=new int[]{+1,0,-1,0};

            for(int d=0;d<4;d++){
                int nrow=row+drow[d];
                int ncol=col+dcol[d];
                if(nrow>=0 && nrow<n && ncol>=0 && ncol<m){
                    if(grid[nrow][ncol]==0 && vis[nrow][ncol][remObs]==0){
                        q.offer(new Pair(1+currSteps,nrow,ncol,remObs));
                        vis[nrow][ncol][remObs]=1;
                    }
                    else if(grid[nrow][ncol]==1 && remObs>0 && vis[nrow][ncol][remObs-1]==0){
                        q.offer(new Pair(1+currSteps,nrow,ncol,remObs-1));
                        vis[nrow][ncol][remObs-1]=1;
                    }
                }
            }
        }
        return -1;
    }
}